File: src/models/sensor.py
from typing import Any, ClassVar, Dict, Mapping, Optional, Sequence, Tuple
from datetime import datetime, timezone

from typing_extensions import Self
from viam.components.sensor import Sensor
from viam.proto.app.robot import ComponentConfig
from viam.proto.common import Geometry, ResourceName
from viam.resource.base import ResourceBase
from viam.resource.easy_resource import EasyResource
from viam.resource.types import Model, ModelFamily
from viam.utils import SensorReading, ValueTypes


class UserActivationSensor(Sensor, EasyResource):
    MODEL: ClassVar[Model] = Model(ModelFamily("viam-soleng", "user-activation"), "sensor")
    
    # Internal state
    activation_date: Optional[str] = None  # ISO 8601 timestamp or None
    delete_flag: bool = False  # Whether account is marked for deletion

    @classmethod
    def new(
        cls, config: ComponentConfig, dependencies: Mapping[ResourceName, ResourceBase]
    ) -> Self:
        """This method creates a new instance of this Sensor component.
        The default implementation sets the name from the `config` parameter and then calls `reconfigure`.

        Args:
            config (ComponentConfig): The configuration for this resource
            dependencies (Mapping[ResourceName, ResourceBase]): The dependencies (both required and optional)

        Returns:
            Self: The resource
        """
        return super().new(config, dependencies)

    @classmethod
    def validate_config(
        cls, config: ComponentConfig
    ) -> Tuple[Sequence[str], Sequence[str]]:
        """This method allows you to validate the configuration object received from the machine,
        as well as to return any required dependencies or optional dependencies based on that `config`.

        Args:
            config (ComponentConfig): The configuration for this resource

        Returns:
            Tuple[Sequence[str], Sequence[str]]: A tuple where the
                first element is a list of required dependencies and the
                second element is a list of optional dependencies
        """
        # No dependencies needed for this sensor
        return [], []

    def reconfigure(
        self, config: ComponentConfig, dependencies: Mapping[ResourceName, ResourceBase]
    ):
        """This method allows you to dynamically update your service when it receives a new `config` object.

        Args:
            config (ComponentConfig): The new configuration
            dependencies (Mapping[ResourceName, ResourceBase]): Any dependencies (both required and optional)
        """
        # Load attributes from config
        attrs = dict(config.attributes.fields)
        
        # Load activation_date (defaults to None)
        if "activation_date" in attrs:
            value = attrs["activation_date"]
            if value.HasField("null_value"):
                self.activation_date = None
            elif value.HasField("string_value"):
                self.activation_date = value.string_value
            else:
                self.activation_date = None
        else:
            self.activation_date = None
        
        # Load delete_flag (defaults to False)
        if "delete_flag" in attrs:
            value = attrs["delete_flag"]
            if value.HasField("bool_value"):
                self.delete_flag = value.bool_value
            else:
                self.delete_flag = False
        else:
            self.delete_flag = False
        
        self.logger.info(
            f"Reconfigured: activation_date={self.activation_date}, delete_flag={self.delete_flag}"
        )

    async def get_readings(
        self,
        *,
        extra: Optional[Mapping[str, Any]] = None,
        timeout: Optional[float] = None,
        **kwargs
    ) -> Mapping[str, SensorReading]:
        """Returns current activation state as sensor readings.
        
        Returns:
            Mapping[str, SensorReading]: Dictionary with activation_date and delete_flag
        """
        return {
            "activation_date": self.activation_date,
            "delete_flag": self.delete_flag
        }

    async def do_command(
        self,
        command: Mapping[str, ValueTypes],
        *,
        timeout: Optional[float] = None,
        **kwargs
    ) -> Mapping[str, ValueTypes]:
        """Handles DoCommand requests for activation state management.
        
        Supported commands:
        - get_activation_state: Returns current state
        
        Args:
            command (Mapping[str, ValueTypes]): The command to execute
            
        Returns:
            Mapping[str, ValueTypes]: Command response
        """
        cmd = command.get("command")
        
        if cmd == "get_activation_state":
            return self._get_activation_state()
        
        else:
            error_msg = f"Unknown command: {cmd}"
            self.logger.error(error_msg)
            return {"success": False, "error": error_msg}

    def _get_activation_state(self) -> Dict[str, Any]:
        """Returns current activation state."""
        self.logger.debug("Getting activation state")
        return {
            "activation_date": self.activation_date,
            "delete_flag": self.delete_flag
        }

    async def get_geometries(
        self, *, extra: Optional[Dict[str, Any]] = None, timeout: Optional[float] = None
    ) -> Sequence[Geometry]:
        self.logger.error("`get_geometries` is not implemented")
        raise NotImplementedError()

File: models/sensor.go
// Package models implements the viam-soleng:user-activation:sensor model.
package models

import (
	"context"
	"sync"

	"go.viam.com/rdk/components/sensor"
	"go.viam.com/rdk/logging"
	"go.viam.com/rdk/resource"
)

// UserActivation is the model triple for this sensor.
var UserActivation = resource.NewModel("viam-soleng", "user-activation", "sensor")

func init() {
	resource.RegisterComponent(sensor.API, UserActivation,
		resource.Registration[sensor.Sensor, *Config]{
			Constructor: newUserActivationSensor,
		},
	)
}

// Config holds the user-configurable attributes for the sensor.
//
// activation_date is an ISO 8601 timestamp (UTC) and may be null/unset, which
// is represented as a nil pointer. delete_flag defaults to false when omitted.
type Config struct {
	ActivationDate *string `json:"activation_date,omitempty"`
	DeleteFlag     bool    `json:"delete_flag,omitempty"`
}

// Validate ensures the config is well-formed and reports dependencies. This
// sensor needs no dependencies, so it returns empty required/optional lists.
func (cfg *Config) Validate(path string) ([]string, []string, error) {
	return nil, nil, nil
}

type userActivationSensor struct {
	resource.Named
	resource.TriviallyCloseable

	logger logging.Logger

	mu             sync.Mutex
	activationDate *string
	deleteFlag     bool
}

func newUserActivationSensor(ctx context.Context, deps resource.Dependencies, conf resource.Config, logger logging.Logger) (sensor.Sensor, error) {
	s := &userActivationSensor{
		Named:  conf.ResourceName().AsNamed(),
		logger: logger,
	}
	if err := s.Reconfigure(ctx, deps, conf); err != nil {
		return nil, err
	}
	return s, nil
}

// Reconfigure dynamically updates the sensor's state when it receives a new
// config object.
func (s *userActivationSensor) Reconfigure(ctx context.Context, deps resource.Dependencies, conf resource.Config) error {
	cfg, err := resource.NativeConfig[*Config](conf)
	if err != nil {
		return err
	}

	s.mu.Lock()
	defer s.mu.Unlock()
	s.activationDate = cfg.ActivationDate
	s.deleteFlag = cfg.DeleteFlag

	s.logger.Infof("Reconfigured: activation_date=%v, delete_flag=%v", activationDateLog(s.activationDate), s.deleteFlag)
	return nil
}

// Readings returns the current activation state as sensor readings.
func (s *userActivationSensor) Readings(ctx context.Context, extra map[string]interface{}) (map[string]interface{}, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.activationState(), nil
}

// DoCommand handles DoCommand requests for activation state management.
//
// Supported commands:
//   - get_activation_state: returns the current state
func (s *userActivationSensor) DoCommand(ctx context.Context, command map[string]interface{}) (map[string]interface{}, error) {
	cmd, _ := command["command"].(string)

	switch cmd {
	case "get_activation_state":
		s.logger.Debug("Getting activation state")
		s.mu.Lock()
		defer s.mu.Unlock()
		return s.activationState(), nil
	default:
		errMsg := "Unknown command: " + cmd
		s.logger.Error(errMsg)
		return map[string]interface{}{"success": false, "error": errMsg}, nil
	}
}

// activationState returns the current state as a map. Callers must hold s.mu.
func (s *userActivationSensor) activationState() map[string]interface{} {
	var activationDate interface{}
	if s.activationDate != nil {
		activationDate = *s.activationDate
	}
	return map[string]interface{}{
		"activation_date": activationDate,
		"delete_flag":     s.deleteFlag,
	}
}

// activationDateLog renders a nil pointer as "<nil>" for logging, otherwise the
// underlying string value.
func activationDateLog(d *string) interface{} {
	if d == nil {
		return nil
	}
	return *d
}
